index-server/main.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/deislabs/oras/pkg/content"
+	"github.com/deislabs/oras/pkg/oras"
+	indexSchema "github.com/devfile/registry-support/index/generator/schema"
+
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/gin-gonic/gin"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+const (
+	devfileName            = "devfile.yaml"
+	devfileConfigMediaType = "application/vnd.devfileio.devfile.config.v2+json"
+	devfileMediaType       = "application/vnd.devfileio.devfile.layer.v1"
+	registryPath           = "/registry/stacks"
+	indexPath              = "/registry/index.json"
+	scheme                 = "http"
+	registryService        = "localhost:5000"
+)
+
+func main() {
+	// Wait until registry is up and running
+	isDone := false
+	for !isDone {
+		resp, err := http.Get(scheme + "://" + registryService)
+		if err != nil {
+			log.Fatal(err.Error())
+		}
+
+		if resp.StatusCode == http.StatusOK {
+			isDone = true
+			log.Println("Registry is up and running")
+		}
+		log.Println("Waiting for registry to start...")
+		time.Sleep(time.Second)
+	}
+
+	// Load index file
+	bytes, err := ioutil.ReadFile(indexPath)
+	if err != nil {
+		log.Fatalf("failed to read index file: %v", err)
+	}
+
+	var index []indexSchema.Schema
+	err = json.Unmarshal(bytes, &index)
+	if err != nil {
+		log.Fatalf("failed to unmarshal index file: %v", err)
+	}
+
+	// Before starting the server, push the devfile artifacts to the registry
+	for _, devfileIndex := range index {
+		err := pushStackToRegistry(devfileIndex)
+		if err != nil {
+			log.Fatal(err.Error())
+		}
+	}
+
+	// Start the server and serve requests and index.json
+	router := gin.Default()
+
+	router.GET("/health", func(c *gin.Context) {
+		c.JSON(http.StatusOK, gin.H{
+			"message": "the server is up and running",
+		})
+	})
+
+	router.GET("/devfiles/:name", func(c *gin.Context) {
+		name := c.Param("name")
+		for _, devfileIndex := range index {
+			if devfileIndex.Name == name {
+				bytes, err := pullStackFromRegistry(devfileIndex)
```

Comment:
Originally I design the way like that is because it's easier to let consumers integrate (eg. odo already implemented HTTP request internally so odo just need to update the endpoint we provide), but since we decide to let consumers use the library we provide for integration, then I think it's fine we remove the endpoint `/devfiles/<name>` and let user browser the whole stacks by serving the whole stacks as static files.

I can give a try to see if Gin API `router.StaticFile()` can serve the whole stacks properly so that user can view the whole stacks instead of only devfile.